### DISCUSSION TITLE
Binding a pandas dataframe to AG Grid

### DISCUSSION THREAD

**Discussion by stdusr** [zauberzeug/nicegui#1713] [Q&A] [answered]
Good day,
is there a way to display pandas dataframe using AG Grid and auto-update it whenever the dataframe changes? The [binding](https://nicegui.io/documentation/bindings) functions doesn't seem to work with AG Grid. I don't have a minimal code example as don't know where to start but would be really grafeful for a hint to the right direction.

**Comment by FabianGoessling** [comment, marked as answer]
Hey, if you look at the documentation it shows how to update the grid 'manually'. So modifying the *aggrid* example:
```python
from nicegui import ui
import pandas as pd
import random

grid = ui.aggrid(
    {
        "defaultColDef": {"flex": 1},
        "columnDefs": [
            {"headerName": "Name", "field": "name"},
            {"headerName": "Age", "field": "age"},
            {"headerName": "Parent", "field": "parent", "hide": True},
        ],
        "rowData": [],
        "rowSelection": "multiple",
    }
).classes("max-h-40")


class App:
    def __init__(self):
        self.rows = []

    def update(self, df):
        self.rows.clear()
        self.rows.extend(df.to_dict("records"))


app = App()
grid.options["rowData"] = app.rows


def update():
    df = pd.DataFrame(
        [
            {"name": "Alice", "age": random.random(), "parent": "David"},
            {"name": "Bob", "age": random.random(), "parent": "Eve"},
            {"name": "Carol", "age": random.random(), "parent": "Frank"},
        ]
    )
    app.update(df)
    grid.update()


ui.button("Update", on_click=update)
ui.run(reload=False)

```
If you want to further unlink the data from your frontend you could split the update function, as it now modifies app (backend) and grid (frontend) at once. You could also think about setting up a loop "on_startup", which checks for diffs in the data and then updates the grid ( see e.g. lifecycle/events part of the official documentation). Also, there is a from_pandas method on the aggrid object, so maybe this is of help.. Note that the above example can be simplified if you are fine with overwriting the grid.options["rowData"] directly.

**Reply by stdusr** [reply]
[FabianGoessling](https://github.com/FabianGoessling), many many thanks for elaborate answer!

**Reply by akai-1024** [reply]
I think use "grid.call_api_method('setRowData', row_data)" for updating data is the correct method.
When I use grid.update(), it will refresh all ui.aggrid widget ui.
